### ISSUE TITLE
Add a new provider called the "MaybeProvider"

### ISSUE THREAD

**Issue by temcewen** [rrousselGit/riverpod#2398]
**Is your feature request related to a problem? Please describe.**
I also describe the problem here: https://github.com/rrousselGit/riverpod/discussions/2388

To summarize: the issue is that data often has multiple sources that shouldn't all be evaluated, may not be known until runtime, and could be the direct result of a widget's build method. Providers are currently limited because they have the relative restriction of always getting data from sources that are known, already evaluated, or they cannot be set during a widget's build method (for good reasons).

The example I gave in the linked discussion forum is the easiest for me to think about. Imagine you have a post model for a social media. This model can come from the API directly via its GUID, but it can also come from a request to a paginator for the feed or from a request to the owning user's profile.

Modeling the autodisposable family provider for this post model could look like this:

```
@riverpod
Future<PostModel> fetchPost(string id) {
   // create api request for the model via the id
   return await dataSource.getPost(id);
}
```

Notice how the provider specifically doesn't use any references to providers for the feed or the profile? It's because doing so is practically useless! Consider that if you were to watch the autodisposable family providers for the post and feed, you wouldn't be able to do so for any specific post model because you first need to have context for them. For the feed paginator provider, you would need to know exactly what page the post model is on which is impossible until runtime. For the profile provider, you would need to know what user the post belongs to which is also impossible until runtime. 

You might be thinking, just create providers for the feedPostFamilyProvider and profilePostFamilyProviders which evaluate to null until the values are set at which point the postFamilyProvider could use them to get the most recent version of the post model.

This would look like:

```
@riverpod
Future<PostModel> fetchPost(string id) {
   // all are nullable values
   var feedPostModel = ref.watch(feedPostProvider(id));
   var profilePostModel = ref.watch(profilePostProvider(id));

   // only fetch the post from the api if it doesn't exist from the feed or profile providers
   var apiPostModel = (feedPostModel  ?? profilePostModel) == null ? await dataSource.getPost(id) : null;

   // get the most recent version of post model and ignore nullable values
   var recentVersion = _mostRecent([feedPostModel, profilePostModel, postAPIProvider]);

   return recentVersion;
}
```

The issue is that in order for feedPostModel and profilePostModel to work, they need to be able to evaluate to null and later on be set to an actual value which will cause postProvider to reevaluate and choose the most recent model.

This means that they should probably be autodisposable family Notifier<PostModel> with build methods that create null values. When they are ready to be set, a method will be called on them which will cause their values to be set. This is a problematic solution because these values won't always be set by a specific user action, but often they should often be set through the build methods of widgets. For example, the feed must be evaluated as soon as the app is opened. This means that it should always be evaluated while a widget is being built.  The user's own profile will also need to be evaluated when the app is built. This means that the notifier's methods should be called while widgets are building which is a bad practice according to Riverpod's best practices: https://github.com/rrousselGit/riverpod/issues/1762

**Describe the solution you'd like**
I think the creation of something called the "MaybeProvider" would not only simplify the way this problem can be solved, but it would also allow for a relatively clean solution that still adheres to a reactive way of thinking. 

The essential idea of a MaybeProvider is that it decouples the observation of a provider from its evaluation. 

The MaybeProvider would use these simple principles:

1. ref.watch never calls it's build method, but always forces an immediate unwrapping of its value which would evaluate to null
2. ref.build must be called on a MaybeProvider for its build function to run
3. ref.build is only initially evaluated once just like ref.watch, and it only reevaluates if one of its dependencies changes (through ref.watch)

The advantage here is that you can defer the evaluation of a Provider while still observing its value. Take the previous example but assume that feedPostProvider and profilePostProvider are autodisposable family MaybeProviders:

```
@riverpod
Future<PostModel> fetchPost(string id) {
   // all are nullable values
   var feedPostModel = ref.watch(feedPostProvider(id));
   var profilePostModel = ref.watch(profilePostProvider(id));
   
   // only fetch the post from the api if it doesn't exist from the feed or profile providers
   var apiPostModel = (feedPostModel  ?? profilePostModel) == null ? await dataSource.getPost(id) : null;

   // get the most recent version of post model and ignore nullable values
   var recentVersion = _mostRecent([feedPostModel, profilePostModel, postAPIProvider]);

   return recentVersion;
}
```

Now, if the particular family values for `feedPostProvider(id)` and `profilePostProvider(id)` haven't had `ref.build` called on them, they will immediately evaluate to null. However, if `ref.build` is called on them in the future, the observation through `ref.watch` will still trigger a rebuild of postProvider. 

`ref.build` should be able to be called from a widget's build method. While this would cause its dependent providers to reevaluate, it won't cause them to do so during the widget's build. Why? Because ref.build of a MaybeProvider _must always be asynchronous_. To further clarify, this is what the implementation of a MaybeProvider could look like for a post model that is retrieved from the profile:

```
@Riverpod(maybe: true)
class FetchProfile extends _$FetchProfile {
   // this results in the generator creating the family version of this provider
   ProfileModel? get(string profileId);

   // arguments after ref are NOT the family input
   @override
   Future<ProfileModel> build(FetchProfilePostRef ref, string profileId) async {
      var profile = await dataSource.getProfile(profileId);
      
      for (var post in profile.posts) {
         // second argument after the provider is the build argument
         await ref.build(fetchProfilePostProvider(post.id), {profile.id, post.id});
      }
      return ref.watch(fetchProfileProvider(arg.profile))[arg.postId];
   }
}

@Riverpod(maybe: true)
class FetchProfilePost extends _$FetchProfilePost {

   // this results in the generator creating the family version of this provider
   PostModel? get(string postId);

   // arguments after ref are NOT the family input
   @override
   Future<PostModel> build(FetchProfilePostRef ref, {required string profileId, required string postId}) async {
      return ref.watch(fetchProfileProvider(profileId))[postId];
   }
}

@riverpod
Future<PostModel> fetchPost(string id) {
   // all are nullable values
   var feedPostModel = ref.watch(fetchFeedPostProvider(id));
   var profilePostModel = ref.watch(fetchProfilePostProvider(id));
   
   // only fetch the post from the api if it doesn't exist from the feed or profile providers
   var apiPostModel = (feedPostModel  ?? profilePostModel) == null ? await dataSource.getPost(id) : null;

   // get the most recent version of post model and ignore nullable values
   var recentVersion = _mostRecent([feedPostModel, profilePostModel, postAPIProvider]);

   return recentVersion;
}
```

I omitted the fetchFeedPostProvider for brevity. It's important to reiterate that the build argument which is provided is NOT the same as the family id. This allows MaybeProviders which aren't families to also have arguments which are provided and it gives the ability for complex connections between providers which won't be known until runtime. 

It's also important to note that ref.build will only be evaluated ONCE and not every time it's called. It's just like all other build methods in that the initial evaluation only happens when it's first called, and in order for it to be reevaluated, it must be due to a change in one of its dependencies. This prevents any sort of circular dependency from happening when build is called on a provider that's dependent on the provider which called build like in the given example of fetchProfileProvider and fetchProfilePostProvider.

TLDR the idea is simply to have a provider which decouples observation from evaluation. My proposal doesn't have to be the way that it's done, but I think it needs to be done somehow because otherwise enterprise level apps with data which comes from multiple sources are going to continue to use wacky hacks or just abandon Riverpod altogether.

**Describe alternatives you've considered**
I gave most of the alternatives in my description of the problem. There is a "hack" that sort of works for now by using a FutureProvider with an asynchronous build method and calling the notifiers of other providers in the build method after it awaits some particular data. While this does work, it's against the best practices. I think there needs to be a recommended solution for this problem. 

One user, @mattermoran, said that they use Hive as a cache and the ultimate source of truth for models which solves this problem. While this may work, it's weird to me that there isn't a native Riverpod solution. I don't like the idea of needing to "hack" Riverpod to get this done. It means loads and loads of extra hours of work for building enterprise level solutions which will almost always have data coming from multiple sources.

**Additional context**
Any feedback on this idea is appreciated. I'm not the best Flutter or Riverpod programmer out there, so I understand if there are issues with what I'm proposing. I imagine that the biggest problem with this solution is the difficulty in its implementation just because it's so much different than any of the other providers.

**Comment by rrousselGit** [comment]
Are you aware that you can use  `ref.exist(provider)` to know at runtime if a provider is mounted or not?  
This could invalidate the problem described – which centers around being unable to check if a provider is initialized without initializing it.

**Comment by rrousselGit** [comment]
Conversely, I don't think a "MaybeProvider" is the correct path for such problem.

Instead we'd likely want to have ref.watch/ref.listen optionally be "weak" (as in don't trigger a build if non-existent).
This would be compatible with all providers rather than be a new kind of provider.

So we could have something like:

```dart
final provider = FutureProvider<Model>((ref) async {
  Future<Model>
});

**Comment by rrousselGit** [comment]
onversely, I don't think a "MaybeProvider" is the correct path for such problem.

Instead, we'd likely want to have ref.watch/ref.listen optionally be "weak" (as in don't trigger a build if non-existent).
This would be compatible with all providers rather than be a new kind of provider.

So we could have something like:

```dart
final dependent = FuturePovider<Model>(...);

final provider = FutureProvider<Model>((ref) async {
  // returns a nullable value
  Future<Model>? value = ref.weakWatch(dependent.future);

  ref.weakListen<AsyncValue>(dependent, (prev, value) {
    // listen to changes but do not trigger a build
  });

});
```

This is most definitely doable. Although it would require quite a bit of work, and for now I'm focusing on stabilizing Riverpod instead.

**Comment by temcewen** [comment]
@rrousselGit I was aware of ref.exists. Ref.exists isn't a good solution for this problem. The issue is that fetchPostProvider HAS to observe BOTH fetchFeedPostProvider and fetchProfilePostProvider regardless of if they exist or not. This comes back to the essential proposal which is a provider that decouples observation from evaluation.

**Comment by rrousselGit** [comment]
> HAS to observe BOTH fetchFeedPostProvider and fetchProfilePostProvider regardless of if they exist or not.

Why?

**Comment by temcewen** [comment]
The reason is complicated. I described the problem in more detail in the original post and the linked discussion at the top. Essentially, you don't want to have to get the post from every possible source. That would require X number of extra unnecessary requests. However, there are times when a new request will be made from a different source and then the original provider should be reevaluated. In the example of the post that can come from the API, the feed paginator, or the profile, you don't know when or what page of the feed or profile the post will come from until runtime. However, when the post just happens to show up in the feed or in a particular profile, you would want the post's value to be updated for all of its dependent widgets/providers. This allows you to do a simple ref.watch(fetchPostProvider) and expect that it will get the most recent data regardless of its source and regardless of the possibilities of what sources may or may not be evaluated yet.

The solution is very similar to what you're saying about a weakWatch, and that could possibly solve the problem as well.

**Comment by temcewen** [comment]
This is actually the exact problem that we're dealing with right now. I have a small team and we're creating a social media, but state management is difficult. It seems like Riverpod is the best solution, but this is the one problem we have with it.

@rrousselGit  Anyways, we will find a workaround either way and this will not stop us from using Riverpod. I really appreciate all the work you've put into this project. It's really great and far superior to the alternatives! Right now, we're all broke college students, but if our project works out, we will definitely be sponsoring you so you can continue to do this amazing work!

**Comment by temcewen** [comment]
I want to reiterate again: this isn't a problem that will show up in small and even medium sized app, but I can't imagine an enterprise-grade app where this problem won't show up. Based upon my original post, it's easy to extrapolate why it would happen for any social media like TikTok, Instagram, Facebook, etc. However, to show a completely different type of example, I think it would also happen with the Uber app if it used Riverpod. Imagine that the Uber app user just got connected with their driver. Most likely, the profile summary information for the driver such as their name and rating will be sent alongside the meta information about the ride itself such as the estimated time of arrival, the car type, etc. It only makes sense for it to be like this to reduce request overhead. Of course, many of the details for the ride will be continually updated via a stream such as the estimated time of arrival and the exact location of the driver. However, the profile of the driver won't be included in the stream because that would be useless, extraneous data that wouldn't change under normal conditions, but when it does just so happen to change, that's where the issues become apparent. Now, imagine that the user taps on the driver's name to pull up their profile, and in the time between when the user tapped on the profile and when the ride information was originally sent, another user gave the driver a 1-star rating which lowered the driver's overall rating. This data would show up in the newly retrieved profile of the user along with whatever other data that the original ride request didn't have since it only got profile summary data. Even though the user can see the drivers new rating from the retrieved profile, there isn't a simple way to have that new data propagated back to the original screen with the way Riverpod currently works. Ref.weakWatch or MaybeProvider would work. I would love either and I think ref.weakWatch could even be more versatile because it can be used on any type of provider.

**Comment by rrousselGit** [comment]
Flagging as planned. I'll use a slightly different API, but the principle is there.

**Comment by temcewen** [comment]
@rrousselGit Thanks so much for this! I'm really excited to try it out! You're one of my favorite open source devs. I wish I could do more, but I'm buying you a coffee for now 😊

**Comment by AlexeyKatsuro** [comment]
Hi @rrousselGit.
I have a question about `ref.listen(provider, weak: true)`. While this listener is alive, will it prevent the provider from being disposed of?.

I'm trying to implement behavior where providers can be tied to the lifecycle of another ref.

I have a HomeScreen, and on it, there's a TabView with nested pages TabPage1, TabPage2. I want to make it so that the tabDataProviders has a lifecycle similar to HomeScreen's, not TabPages.

Currently, I do 
```dart
// Inside HomeNotifier
homeRef.listen(tab1DataProvider, doNothing); 
homeRef.listen(tab2DataProvider, doNothing) 
```
to keep them alive while HomeScreen it is. 

But I lose the "lazy" initialization. And tab2DataProvider will be launched even if the user has not opened TabPage2. 
I want to tie them to home to prevent disposing when the TabPage1/2 widget is destroyed. I would be very grateful if you could suggest how I might achieve this?

P.S. I understand that this example could be solved at the widget lifecycle level. My actual use case can't avoid widget disposing. Therefore, I can rely on the lifespan of HomeNotifier.

**Comment by temcewen** [comment]
@AlexeyKatsuro If I'm understanding correctly, why can't you do a lazy listen to the tab providers in the home notifier, but do a "watch" in the widgets. When the widgets of the tab providers are disposed, the corresponding tab providers would be disposed as well. However, you could set the tab providers to be "keepAlive: true" in your Riverpod annotation. This would make it so the widgets are disposed, the home notifier can lazily watch the tab providers, and the tab providers will stay alive even after their widgets are disposed. Would that solve your problem?

**Comment by AlexeyKatsuro** [comment]
@temcewen, because Riverpod doesn't have a listener(lazy: true) or something similar. As soon as HomeNotifier is created, it will call .listen(tab2DataProvider, doNothing) to link the lifecycle, thereby creating the provider.

Regarding `keepAlive: true`, it also cannot be a solution because I want them to be cleared after HomeNotifier is disposed of.

**Comment by temcewen** [comment]
This whole thread is about creating a lazy listener, and Remi has already written the code afaik. I don't think it's reached a release mode yet, but rest assured, it's coming!

**Comment by rrousselGit** [comment]
> I have a question about ref.listen(provider, weak: true). While this listener is alive, will it prevent the provider from being disposed of?.

No.  
`ref.listen(p, weak: true, ...)` will not prevent providers from getting disposed.

**Comment by AlexeyKatsuro** [comment]
> > I have a question about ref.listen(provider, weak: true). While this listener is alive, will it prevent the provider from being disposed of?.
> 
> No. `ref.listen(p, weak: true, ...)` will not prevent providers from getting disposed.


Oh, judging by the code from the PR, I expected this. Then, `weak: true` won't help me solve my lifecycle issue.
@rrousselGit Do you have any ideas on how to implement behaviour what I described? _(create a subscription but not trigger the provider)_

**Comment by rrousselGit** [comment]
We could always add a `listen(lazy: true)` with that behaviour. Although for now I'm not sure that's worth it.
